src/server.py

```diff
@@ -148,6 +156,38 @@ def binify(bins, data):
     return result
 
 
+def valve(head_rev, pushlog_id, branch, priority):
+    """Determine which kind of job should been returned"""
+    priority = priority
+    BRANCH_COUNTER.increase_the_counter(branch)
+    request_list = []
+    try:
+        request_list = session.query(TaskRequests.head_rev, TaskRequests.pushlog_id,
+                                     TaskRequests.priority).limit(40)
+    except Exception:
+        session.rollback()
+
+    requests = {}
+    for head_rev, pushlog_id, priority in request_list:
+        requests[pushlog_id] = {'head_rev': head_rev,
+                                'priority': priority}
+
+    # If this pushlog_id has been schduled, we just return
+    # the priority returned before.
+    if pushlog_id in requests.keys():
+        priority = requests.get(pushlog_id)['priority']
+    else:
+
+        # we return all jobs for every 5 pushes.
+        if RequestCounter.BRANCH_COUNTER[branch] >= 5:
+            RequestCounter.reset(branch)
+            priority = None
+        task_request = TaskRequests(str(head_rev), str(pushlog_id), priority)
+        session.add(task_request)
+        session.commit()
```

Comment:
I do question if we need RequestCounter and TaskRequests- I would believe we can get by just fine with TaskRequests since pushlog_id will be branch specific- lets add a field 'branch' to the TaskRequests, and we should be able to simplify.  Now if we request data by pushlog_id, then it won't work- I think we need to think about query speed and determine if two small queries is better than 1 sort of small query.  Also for code maintainability- overall this two table structure I don't like, but it might be most efficient if done right.